### ISSUE TITLE
Unable to use enum as array index.

### ISSUE THREAD

**Issue by sktt** [bpftrace/bpftrace#2701]
### What reproduces the bug?
```sh
sudo bpftrace -e 'enum foo { BAR }; struct MyStruct { int y[4]; } BEGIN {printf("%d\n", BAR); $s = (struct MyStruct *) 0;  $s->y[BAR]; }'
```
Results in
```
stdin:1:107-117: ERROR: The array index operator [] only accepts literal integer indices.
```

Subsituting `BAR` for `0`
```sh
 sudo bpftrace -e 'enum foo { BAR }; struct MyStruct { int y[4]; } BEGIN {printf("%d\n", BAR); $s = (struct MyStruct *) 0;  $s->y[0]; }'
 ```
shows the expected result:
 ```
WARNING: Addrspace is not set
Attaching 1 probe...
0 
```

### bpftrace --info

```
System
  OS: Linux 6.1.0-9-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.27-1 (2023-05-08)
  Arch: x86_64

Build
  version: v0.17.0
  LLVM: 14.0.6
  unsafe uprobe: no
  bfd: no
  libdw (DWARF support): yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes
  skboutput: no

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes
  kprobe_multi: no
  raw_tp_special: yes

```

**Comment by danobi** [comment]
Ah yeah that's cuz `BAR` is not an integer literal -- it's an identifier. bpftrace can/should do a better job of constant folding.

**Comment by sktt** [comment]
Thanks for pointing out that this should be fixed by a constant folding pass!

**Comment by viktormalik** [comment]
> Thanks for pointing out that this should be fixed by a constant folding pass!

Yes, but using LLVM's constant folding is not going to work. The enum types are not a part of the LLVM program itself, we're storing them separately in [BPFtrace::enums_](https://github.com/iovisor/bpftrace/blob/master/src/bpftrace.h#L181). So, we either need to write our own constant folding pass or to resolve them directly during semantic analysis of the array accesss operator. If this problem only applies to array indexing, the latter could be sufficient (and simpler).

**Comment by ZelinMa557** [comment]
Maybe I can work on fix the bug?

**Comment by danobi** [comment]
It'd be nice to have a generic pass that could collapse const idents -> const. That way it'll be consistent across all our language constructs.

@ZelinMa557 go for it!

**Comment by ZelinMa557** [comment]
> It'd be nice to have a generic pass that could collapse const idents -> const. That way it'll be consistent across all our language constructs.
> 
> @ZelinMa557 go for it!

Thank you. I want to know that is there any other const idents besides enums?

**Comment by danobi** [comment]
I think code like:

```
$a = 5;
@[a] = 3;
```
could be folded into:

```
@[5] = 3;
```

Similarly something like:
```
$a = 2 + 3;
@[a] = 3;
```

could be folded into:
```
@[5] = 3;
```

But I'm not sure we would see much advantage in doing that cuz llvm probably does it for us already. So unlikely to affect generated code.

Maybe keep it simple and start with folding enum idents first? That has clear practical benefits. The constant folding pass can always be extended to do more over time.

**Comment by ZelinMa557** [comment]
> I think code like:
> 
> ```
> $a = 5;
> @[a] = 3;
> ```
> 
> could be folded into:
> 
> ```
> @[5] = 3;
> ```
> 
> Similarly something like:
> 
> ```
> $a = 2 + 3;
> @[a] = 3;
> ```
> 
> could be folded into:
> 
> ```
> @[5] = 3;
> ```
> 
> But I'm not sure we would see much advantage in doing that cuz llvm probably does it for us already. So unlikely to affect generated code.
> 
> Maybe keep it simple and start with folding enum idents first? That has clear practical benefits. The constant folding pass can always be extended to do more over time.

Thanks, I decide to start with folding enum idents & binary expressions. Also I hope this change can be limited to the scope of ast and not affect semantic analysis and code generation, so I had best modify the AST during the constant fold pass. For example, `1+2+3` will be replaced by `6` before semantic analysis & codegen. Do you think my solution is reasonable, or that AST should not be modified during constant fold pass?

**Comment by danobi** [comment]
Yeah, that makes sense. Constant folding should probably occur before semantic analysis, at least to solve the issue in the original post.

**Comment by fbs** [comment]
There is a Mutator visitor that *should* work for a mutating pass, but its never been really used before and might have some rough edges

**Comment by jordalgo** [comment]
This now works as expected:
```
$ sudo bpftrace -e 'enum foo { BAR }; struct MyStruct { int y[4]; } BEGIN {printf("%d\n", BAR); $s = (struct MyStruct *) 0;  $s->y[BAR]; }'

Attached 1 probe
0
```
